packages/firestore/src/model/transform_operation.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FieldValue, ServerTimestampValue, ArrayValue } from './field_value';
+import { Timestamp } from '../api/timestamp';
+import * as misc from '../util/misc';
+import { assert } from '../util/assert';
+
+/** Represents a transform within a TransformMutation. */
+export interface TransformOperation {
+  /** Transforms the provided `previousValue`. */
+  transform(previousValue: FieldValue, localWriteTime?: Timestamp): FieldValue;
+
+  isEqual(other: TransformOperation): boolean;
+}
+
+/** Transforms a value into a server-generated timestamp. */
+export class ServerTimestampTransform implements TransformOperation {
+  private constructor() {}
+  static instance = new ServerTimestampTransform();
+
+  transform(previousValue: FieldValue, localWriteTime?: Timestamp): FieldValue {
```

Comment:
If you end up following the changes linked to above, you could take the `serverTransformResults` as an argument (you might also have to split this up into `.localTransform` and `.serverTransform`).

With that: 
- You can just call `transform.transform()` in mutation.ts.
- You can assert that the server transform value in null in `ArrayUnion/RemoveTransformOperation` .

Reply:
Though I didn't do the above changes, I did go ahead and change this as it's something I had already been considering. It added more lines of code, but is ultimately cleaner, I think.

NOTE: I did not add the assert for the transform value being null, since I don't want to prevent the server from changing that behavior in the future. We may decide that the server really should send a transformResult and we don't want that to break old clients.  In general, we should assert about our own state, but be careful about asserting on what the server sends us.

Reply:
Thanks, I do think this interaction is now much cleaner and much more obvious.

(... and I would hope that the backend would rev their protocol version before changing the semantics of a response).